File: py-server/utils/dimensions.py
"""
Dimension utilities for Pixora.
"""
from typing import Tuple, Optional


def compute_dimensions(
    original_width: int,
    original_height: int,
    target_width: Optional[int],
    target_height: Optional[int],
    preserve_aspect: bool,
) -> Tuple[int, int]:
    """
    Compute output dimensions given constraints.
    Returns (width, height).
    """
    if not target_width and not target_height:
        return original_width, original_height

    if preserve_aspect:
        if target_width and target_height:
            ratio = min(target_width / original_width, target_height / original_height)
            return max(1, int(original_width * ratio)), max(1, int(original_height * ratio))
        elif target_width:
            ratio = target_width / original_width
            return target_width, max(1, int(original_height * ratio))
        else:
            ratio = target_height / original_height
            return max(1, int(original_width * ratio)), target_height
    else:
        w = target_width if target_width else original_width
        h = target_height if target_height else original_height
        return w, h

File: src-tauri/src/utils/dimensions.rs
/// Dimension computation for resizing.

/// Compute output (width, height) given optional targets and aspect-lock.
pub fn compute_dimensions(
    orig_w: u32,
    orig_h: u32,
    target_w: Option<u32>,
    target_h: Option<u32>,
    preserve_aspect: bool,
) -> (u32, u32) {
    match (target_w, target_h) {
        (None, None) => (orig_w, orig_h),

        (Some(w), Some(h)) if preserve_aspect => {
            let ratio_w = w as f64 / orig_w as f64;
            let ratio_h = h as f64 / orig_h as f64;
            let ratio   = ratio_w.min(ratio_h);
            (
                ((orig_w as f64 * ratio).round() as u32).max(1),
                ((orig_h as f64 * ratio).round() as u32).max(1),
            )
        }

        (Some(w), Some(h)) => (w, h),

        (Some(w), None) if preserve_aspect => {
            let ratio = w as f64 / orig_w as f64;
            (w, ((orig_h as f64 * ratio).round() as u32).max(1))
        }

        (None, Some(h)) if preserve_aspect => {
            let ratio = h as f64 / orig_h as f64;
            (((orig_w as f64 * ratio).round() as u32).max(1), h)
        }

        (Some(w), None)  => (w, orig_h),
        (None,  Some(h)) => (orig_w, h),
    }
}